_posts/2021-09-17-the-joy-of-slow-programming.md

```diff
@@ -0,0 +1,157 @@
+the implementation, added tests soon, but not immediately, and configured CI
+after I had a running version in production.
+
+I was working at a slow, steady pace, trying to enjoy the every moment of it.
+As I removed the pressure of time from myself, I could actually enjoy the
+little details like a [new Rubocop rule that I didn't know before](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Lint/ConstantDefinitionInBlock)
+or read up on [libraries that I was considering to use](https://github.com/bblimke/webmock).
+
+After a few days of working I had the first version and could actually start
+sending notifications to my friend. This meant that I was receiving feedback
+early and could adjust the application based on it, without making too many
+false assumptions.
+
+Was the fast "hacking evening" with Python a wasted time? No, not really.
+Thanks to this session I learned that Telegram is a messaging app where it's
+extremely easy to send messages programmatically and how to create my own
+Telegram bot. From the time perspective I think about it as prototyping session,
+although during the actual session I didn't have this prototyping mindset.
+
+### Next boring steps
+
+The scraper was working in production (I quickly set up Heroku) and my friend was
+receiving new apartment offers every day. The next step was very boring - add
+support for more companies. It was a tedious and repetitive activity to write
+this code, but I was never doing it for more than an hour per day and at the
+same time immediately getting the business value out of it - more and more
+apartment offers were reaching my friend - and this kept me motivated.
+
+I always started by copy pasting the previous scraper and the tests for it.
+The code was repetitive but I avoided reaching for any early abstractions.
+The classes were simple and could fit in a single screen. The actual business
+logic - getting information from the HTML using Nokogiri and regular expressions
+- was unique to each company website; with very few common elements.
+
+This approach paid off as soon as I found the first website loading data
+dynamically from an unsecured JSON API :) Because of the lack of any
+authentication it was just easier to access the JSON API directly. My
+abstraction was lightweight so I didn't have to modify it. I just wrote this
+one particular scraper differently.
+
+At some point I introduced the Apartment model to the application. At this
+stage I wasn't yet sure what data I can actually get in a consistent way. The
+only thing that I knew for sure was that I need to store a unique identifier for
+each apartment to avoid sending duplicated notifications after each subsequent
+check. That's why I added a mandatory `external_id` field. Everything else ended
+up in a JSONB `properties` field. This was simple and allowed me to move fast
+with delivering business value while moving slow with technical assumptions.
```

Comment:
```suggestion
with delivering value while moving slow with technical assumptions.
```
I know what you mean by "business value", just wouldn't use it here since it's nothing business-related, and too many things in this world are business-related :)